download_plugins.py

```diff
@@ -1,85 +1,160 @@
+        download_url = "{url}/latest/{plugin}.hpi".format(
+            url=plugin_base_url, plugin=plugin
+        )
     else:
-        download_url = plugin_base_url + "/download/plugins/%(plugin)s/%(version)s/%(plugin)s.hpi" % {'plugin': plugin, 'version': version}
+        download_url = (plugin_base_url +
+                        "/download/plugins/" +
+                        "{plugin}/{version}/{plugin}.hpi".format(
+                            plugin=plugin, version=version
+                        ))
 
     plugin_download = requests.get(download_url, stream=True)
 
-    if plugin_download.status_code is 200:
-        destination_path = download_directory + "/" + plugin + ".hpi"
+    if plugin_download.status_code == 200:
+        destination_path = "{dir}/{plugin}.hpi".format(
+            dir=download_directory, plugin=plugin
+        )
 
-        with open(destination_path, 'wb') as data:
+        with open(destination_path, "wb") as data:
             for chunk in plugin_download.iter_content(chunk_size=128):
                 data.write(chunk)
 
-        print "Downloaded", plugin
-        
+        print("Downloaded {}".format(plugin))
+
     else:
-        print "Error downloading %s. Response:" % plugin
-        print plugin_download.text
+        print("Error downloading {plugin}. Response:\n{response}".format(
+            plugin=plugin, response=plugin_download.text
+        ))
+        global exit_code
         exit_code = 1
 
-    installed_plugins.append(plugin)
 
 # Install each plugin in the supplied json file along with dependencies.
 def install_plugins():
     for plugin, version in plugins.items():
-        print "**** Install Plugin: %s ****" % plugin
+        print("\n**** Install Plugin: {} *****".format(plugin))
         download_plugin(plugin, version)
         get_dependencies(plugin)
 
+    print("\nDownloading all dependencies")
+    print("*****************************************")
+    for plugin, version in stored_plugins.items():
+        download_plugin(plugin, version)
+
+    for dependency, info in dependencies_info.items():
+        if not (info["duplicate"]):
+            continue
+        print(dedent("""\
+        Warning: Multiple versions of {} found in dependencies.\
+        """.format(dependency)))
+        for parent in info["parents"]:
+            for version, plugin in parent.items():
+                print("    Version {version} required by {plugin}".format(
+                    version=version, plugin=plugin
+                ))
+        print("    Downloaded version {}".format(stored_plugins[dependency]))
+
+
+# Insert a new dependency parent sorted by version
+def version_sorted_insert(dependency, plugin):
+    for idx, elem in enumerate(
+            dependencies_info[dependency["name"]]["parents"]):
+        for version, parent in elem.items():
+            if (LooseVersion(dependency["version"]) <=
+                    LooseVersion(version)):
```

Comment:
Just my two cents, since using `LooseVersion` to compare versions in lots of places and it usually cause the codes too long and it also forces us to break the line, it probably is worth to have a function to handle this logic, for example:

```python
# check if version_1 is older than version_2
def is_older(version_1, version2):
    return LooseVersion(version_1) < LooseVersion(version2)
```
and call the function like:

```python
if id_older(dependency["version"], version):
    do_something()
```